### ISSUE TITLE
Add Spec.CloudProvider in KubeadmConfig

### ISSUE THREAD

**Issue by vincepri** [kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm#157]
/kind feature

**Describe the solution you'd like**
An optional field in the KubeadmConfig Spec called `CloudProvider` could be useful to implement default logic to set the `cloud-provider` flags in multiple places.

This is an example of a configuration to setup the first controlplane instance:
```
apiVersion: bootstrap.cluster.x-k8s.io/v1alpha2
kind: KubeadmConfig
metadata:
  name: ${CLUSTER_NAME}-controlplane-0
spec:
  initConfiguration:
    nodeRegistration:
      name: '{{ ds.meta_data.hostname }}'
      kubeletExtraArgs:
        cloud-provider: aws
  clusterConfiguration:
    apiServer:
      extraArgs:
        cloud-provider: aws
    controllerManager:
      extraArgs:
        cloud-provider: aws
    certificatesDir: /etc/kubernetes/pki
    controlPlaneEndpoint: ""
    clusterName: ${CLUSTER_NAME}
    dns:
      type: ""
    imageRepository: ""
    kubernetesVersion: v1.14.4
    etcd: {}
    networking:
      dnsDomain: "cluster.local"
      podSubnet: "192.168.0.0/16"
      serviceSubnet: "172.16.0.0/12"
```

As you can see, the `cloud-provider: aws` option is repeated in multiple places. Bringing that up to a unique place (even in Cluster API, if we think it might benefit other bootstrap providers as well) would be beneficial to allow for better defaulting and simplifying the UX.

**Comment by detiber** [comment]
I'm not necessarily sure this is worth doing, since we said that we want to require external cloud providers for beta.

**Comment by vincepri** [comment]
Wouldn't we have to set the `cloud-provider` flag to `external` as well? When we're set to migrate to the external cloud providers, we'll default the proposed field, `Spec.CloudProvider` to `external` but still allow users to overwrite it.

**Comment by detiber** [comment]
@vincepri I'm not sure we would allow users to override it at that point.

One thing we discussed previously, which might make more sense than singling out cloud provider here was to let the infrastructure provider specify additional configuration that is merged with the kubeadm config.

**Comment by vincepri** [comment]
That makes sense, I don't mind either way, after #156 the configuration is looking much better. This is definitely a nice to have 😊, we can always drop it after v1alpha2.
